### PR TITLE
Add months_between Spark function

### LINKED ISSUE #4858: Implement months_between() Spark function
### Description

Implement months_between() Spark function

### PR DESCRIPTION
**Pull request by kerwin-zk**
Fixes #4858 

Spark's implementation:https://github.com/apache/spark/blob/5b609598503df603cbddd5e1adf8d2cb28a5f977/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L310

### CODE REVIEW

**Comment by kerwin-zk** [comment]
@mbasmanova  Could you help review this fuction?

**Comment by kerwin-zk** [comment]
@mbasmanova  Could you help review this fuction?

File: velox/docs/functions/spark/datetime.rst [hunk 1]
```diff
@@ -7,6 +7,10 @@ Convenience Extraction Functions
 
 These functions support TIMESTAMP and DATE input types.
 
+.. spark:function:: months_between(timestamp, timestamp, boolean) -> double
```
Comment (mbasmanova) [inline velox/docs/functions/spark/datetime.rst:10]: Please, move this function to preserve alphabetical order.

**Review by mbasmanova — COMMENTED** [review]
@bikramSingh91 Bikram, would you help review this PR?

File: velox/docs/functions/spark/datetime.rst (hunk 1 above)
Comment (kerwin-zk) [reply velox/docs/functions/spark/datetime.rst]: @mbasmanova done.

**Comment by kerwin-zk** [comment]
@bikramSingh91  Could you help review this fuction?

File: velox/docs/functions/spark/datetime.rst [hunk 2]
```diff
@@ -22,6 +22,10 @@ These functions support TIMESTAMP and DATE input types.
     ``day`` need to be from 1 to 31, and matches the number of days in each month.
     days of ``year-month-day - 1970-01-01`` need to be in the range of INTEGER type.
 
+.. spark:function:: months_between(timestamp, timestamp, boolean) -> double
```
Comment (jinchengchenghh) [inline velox/docs/functions/spark/datetime.rst:25]: months_between(timestamp1, timestamp2, boolean)

Comment (jinchengchenghh) [reply velox/docs/functions/spark/datetime.rst:25]: What's the meaning of `boolean`?

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 3]
```diff
@@ -209,4 +209,93 @@ struct LastDayFunction {
...
 
+template <typename T>
+struct MonthsBetweenFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<Timestamp>* /*timestamp1*/,
+      const arg_type<Timestamp>* /*timestamp2*/,
+      const bool /*roundOff*/) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/DateTimeFunctions.h:220]: drop const for `bool`

**Comment by jinchengchenghh** [comment]
Can you use this utility function to simplify the implementation?https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/DateTimeImpl.h#L189

**Comment by jinchengchenghh** [comment]
Paste the Spark implementation to PR description

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 4]
```diff
@@ -208,5 +208,53 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
...
+TEST_F(DateTimeFunctionsTest, monthsBetween) {
+  const auto monthsBetweenFunc = [&](const std::optional<Timestamp> timestamp1,
+                                     const std::optional<Timestamp> timestamp2,
+                                     const std::optional<bool> roundOff) {
+    return evaluateOnce<double>(
+        "months_between(c0, c1, c2)", timestamp1, timestamp2, roundOff);
+  };
+
+  std::string zone = "America/Los_Angeles";
+  using util::fromTimestampString;
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:220]: Used only twice, no need to use using`

File: velox/docs/functions/spark/datetime.rst [hunk 5]
```diff
@@ -22,6 +22,10 @@ These functions support TIMESTAMP and DATE input types.
...
+
+    Returns number of months between times `timestamp1` and `timestamp2`.
```
Comment (rui-mo) [inline velox/docs/functions/spark/datetime.rst:27]: Pleae use format like below for variables.
\`\`day\`\`

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 6]
```diff
@@ -209,4 +209,93 @@ struct LastDayFunction {
...
+  FOLLY_ALWAYS_INLINE int64_t getMonth(const std::tm& time) {
+    return 1 + time.tm_mon;
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t getDay(const std::tm& time) {
+    return time.tm_mday;
+  }
+
+  FOLLY_ALWAYS_INLINE std::tuple<int32_t, int32_t, int32_t, int32_t> splitDate(
+      Timestamp timestamp) {
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:237]: Mabye rename to splitTimestamp to align with the parameter.

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 7]
```diff
@@ -209,4 +209,93 @@ struct LastDayFunction {
...
+    std::tuple<int32_t, int32_t, int32_t, int32_t> result;
+    int32_t year = getYear(getDateTime(timestamp, this->timeZone_));
+    int32_t monthInYear = getMonth(getDateTime(timestamp, this->timeZone_));
+    int32_t dayInMonth = getDay(getDateTime(timestamp, this->timeZone_));
+    int32_t daysToMonthEnd = util::getMaxDayOfMonth(year, monthInYear) -
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:242]: Maybe remove these variables and create the tuple directly because they are used only once.

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 8]
```diff
@@ -209,4 +209,93 @@ struct LastDayFunction {
...
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t getSecondsInDay(
+      Timestamp timestamp,
+      int32_t year,
+      int32_t month,
+      int32_t day) {
+    const date::time_zone* timeZone = this->timeZone_;
+    Timestamp t1 =
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:254]: Maybe t1 -> t or ts

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 9]
```diff
@@ -209,4 +209,93 @@ struct LastDayFunction {
...
+        util::fromDatetime(util::daysSinceEpochFromDate(year, month, day), 0);
+    if (timeZone != nullptr) {
+      t1.toGMT(*timeZone);
+    }
+    return (timestamp.toMicros() - t1.toMicros()) / 1000000;
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:259]: Maybe use existing static variables in Timestamp.h to replace 1000000.
E.g., kMicrosecondsInMillisecond * kMillisecondsInSecond.

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 10]
```diff
@@ -209,4 +209,93 @@ struct LastDayFunction {
...
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      double& result,
+      const arg_type<Timestamp>& timestamp1,
+      const arg_type<Timestamp>& timestamp2,
+      const bool roundOff) {
+    int32_t year1, monthInYear1, dayInMonth1, daysToMonthEnd1;
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:267]: Please follow the code style to declare only one variable in one line.
https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#variables

File: velox/docs/functions/spark/datetime.rst (hunk 2 above)
Comment (kerwin-zk) [reply velox/docs/functions/spark/datetime.rst:25]: @jinchengchenghh  this is the roundOff parameter of spark months_between 
```  
def monthsBetween(
      micros1: Long,
      micros2: Long,
      roundOff: Boolean,
      zoneId: ZoneId): Double = {
    val date1 = microsToDays(micros1, zoneId)
    val date2 = microsToDays(micros2, zoneId)
    val (year1, monthInYear1, dayInMonth1, daysToMonthEnd1) = splitDate(date1)
    val (year2, monthInYear2, dayInMonth2, daysToMonthEnd2) = splitDate(date2)

    val months1 = year1 * 12 + monthInYear1
    val months2 = year2 * 12 + monthInYear2

    val monthDiff = (months1 - months2).toDouble

    if (dayInMonth1 == dayInMonth2 || ((daysToMonthEnd1 == 0) && (daysToMonthEnd2 == 0))) {
      return monthDiff
    }
    // using milliseconds can cause precision loss with more than 8 digits
    // we follow Hive's implementation which uses seconds
    val secondsInDay1 = MICROSECONDS.toSeconds(micros1 - daysToMicros(date1, zoneId))
    val secondsInDay2 = MICROSECONDS.toSeconds(micros2 - daysToMicros(date2, zoneId))
    val secondsDiff = (dayInMonth1 - dayInMonth2) * SECONDS_PER_DAY + secondsInDay1 - secondsInDay2
    val secondsInMonth = DAYS.toSeconds(31)
    val diff = monthDiff + secondsDiff / secondsInMonth.toDouble
    if (roundOff) {
      // rounding to 8 digits
      math.round(diff * 1e8) / 1e8
    } else {
      diff
    }
  }
```

**Comment by kerwin-zk** [comment]
> Can you use this utility function to simplify the implementation?https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/DateTimeImpl.h#L189

@jinchengchenghh This method cannot be used because the month_diff it returns is int, but in spark it is required to be double.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp (hunk 4 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp]: done

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 8 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/DateTimeFunctions.h]: done

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 9 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/DateTimeFunctions.h]: done

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 10 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/DateTimeFunctions.h]: done

File: velox/docs/functions/spark/datetime.rst (hunk 2 above)
Comment (jinchengchenghh) [reply velox/docs/functions/spark/datetime.rst:64]: Please also document here.

**Comment by jinchengchenghh** [comment]
Just the Spark implementation link in PR description is enough, please use the released branch, not the master branch.

File: velox/docs/functions/spark/datetime.rst [hunk 11]
```diff
@@ -61,6 +61,10 @@ These functions support TIMESTAMP and DATE input types.
     ``day`` need to be from 1 to 31, and matches the number of days in each month.
     days of ``year-month-day - 1970-01-01`` need to be in the range of INTEGER type.
 
+.. spark:function:: months_between(timestamp, timestamp, boolean) -> double
```
Comment (rui-mo) [inline velox/docs/functions/spark/datetime.rst:64]: Please distinguish the parameters and use the their names in the content.

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 12]
```diff
@@ -319,4 +319,99 @@ struct DayOfWeekFunction : public InitSessionTimezone<T>,
...
+        ((daysToMonthEnd1 == 0) && (daysToMonthEnd2 == 0))) {
+      result = monthDiff;
+    } else {
+      int64_t secondsInDay1 =
+          getSecondsInDay(timestamp1, year1, monthInYear1, dayInMonth1);
+      int64_t secondsInDay2 =
+          getSecondsInDay(timestamp2, year2, monthInYear2, dayInMonth2);
+      int64_t secondsDiff = (dayInMonth1 - dayInMonth2) * kSecondsInDay +
+          secondsInDay1 - secondsInDay2;
+      int64_t secondsInMonth = 31 * kSecondsInDay;
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:405]: Do we need to consider a month not containing 31 days?

**Review by rui-mo — COMMENTED** [review]
Thanks!

**Comment by kerwin-zk** [comment]
> Just the Spark implementation link in PR description is enough, please use the released branch, not the master branch.

@jinchengchenghh [monthsBetween](https://github.com/apache/spark/blob/5b609598503df603cbddd5e1adf8d2cb28a5f977/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L310)

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 12 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/DateTimeFunctions.h:405]: @rui-mo This part is consistent with the logic of spark.
https://github.com/apache/spark/blob/5b609598503df603cbddd5e1adf8d2cb28a5f977/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L333

File: velox/docs/functions/spark/datetime.rst (hunk 11 above)
Comment (kerwin-zk) [reply velox/docs/functions/spark/datetime.rst]: done

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 6 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/DateTimeFunctions.h]: done

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 7 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/DateTimeFunctions.h]: done

**Comment by kerwin-zk** [comment]
> @kerwin-zk This PR is marked as stale. Could you help rebase and update it? Thanks!

@rui-mo done

File: velox/docs/functions/spark/datetime.rst [hunk 13]
```diff
@@ -85,6 +85,15 @@ These functions support TIMESTAMP and DATE input types.
     ``day`` need to be from 1 to 31, and matches the number of days in each month.
     days of ``year-month-day - 1970-01-01`` need to be in the range of INTEGER type.
 
+.. spark:function:: months_between(timestamp1, timestamp2, roundOff) -> double
+
+    Returns number of months between times ``timestamp1`` and ``timestamp2``.
+    If ``timestamp1`` is later than ``timestamp2``, the result is positive.
+    If ``timestamp1`` and ``timestamp2`` are on the same day of month,
+    or both are the last day of month, returns, time of day will be ignored.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/datetime.rst:93]: typos in "returns, time of day will be ignored."

The description is a bit hard to follow. Could you add some examples?

File: velox/docs/functions/spark/datetime.rst [hunk 14]
```diff
@@ -85,6 +85,15 @@ These functions support TIMESTAMP and DATE input types.
...
+    Otherwise, the difference is calculated based on 31 days per month.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/datetime.rst:94]: What does this mean exactly? Would you add some examples?

File: velox/docs/functions/spark/datetime.rst [hunk 15]
```diff
@@ -85,6 +85,15 @@ These functions support TIMESTAMP and DATE input types.
...
+    The result is rounded to 8 decimal places if ``roundOff`` is set to true.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/datetime.rst:95]: is this '8'? What does that mean? An example would be helpful. What happens if roundOff is false?

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 16]
```diff
@@ -408,4 +408,88 @@ struct DayOfYearFunction {
...
+      bool /*roundOff*/) {
+    timeZone_ = getTimeZoneFromConfig(config);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      double& result,
+      const arg_type<Timestamp>& timestamp1,
+      const arg_type<Timestamp>& timestamp2,
+      bool roundOff) {
+    int32_t year1{0};
```
Comment (mbasmanova) [inline velox/functions/sparksql/DateTimeFunctions.h:429]: These names are hard to read. Consider introducing a struct wth 4 fields.

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 17]
```diff
@@ -408,4 +408,88 @@ struct DayOfYearFunction {
...
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<Timestamp>* /*timestamp1*/,
+      const arg_type<Timestamp>* /*timestamp2*/,
+      bool /*roundOff*/) {
+    timeZone_ = getTimeZoneFromConfig(config);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
```
Comment (mbasmanova) [inline velox/functions/sparksql/DateTimeFunctions.h:424]: Remove FOLLY_ALWAYS_INLINE. This function is quite long.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 18]
```diff
@@ -456,5 +456,52 @@ TEST_F(DateTimeFunctionsTest, quarterDate) {
...
 
+TEST_F(DateTimeFunctionsTest, monthsBetween) {
+  const auto monthsBetweenFunc = [&](const std::optional<Timestamp> timestamp1,
+                                     const std::optional<Timestamp> timestamp2,
+                                     const std::optional<bool> roundOff) {
+    return evaluateOnce<double>(
+        "months_between(c0, c1, c2)", timestamp1, timestamp2, roundOff);
+  };
+
+  std::string zone = "America/Los_Angeles";
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:467]: - zone -> timeZone
- const

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 19]
```diff
@@ -456,5 +456,52 @@ TEST_F(DateTimeFunctionsTest, quarterDate) {
...
+
+  setQueryTimeZone(zone);
+
+  auto monthsBetween = [&](const auto& inputDate1,
+                           const auto& inputDate2,
+                           const auto& roundOff) {
+    Timestamp timestamp1 = util::fromTimestampString(inputDate1);
+    timestamp1.toGMT(util::getTimeZoneID(zone));
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:475]: perhaps, extract a variable to store util::getTimeZoneID(zone)

**Review by mbasmanova — COMMENTED** [review]
Some initial questions and comments.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp (hunk 19 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp]: @mbasmanova done

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp (hunk 18 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp]: @mbasmanova done

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 16 above)
Comment (kerwin-zk) [reply velox/functions/sparksql/DateTimeFunctions.h]: @mbasmanova done

File: velox/docs/functions/spark/datetime.rst (hunk 15 above)
Comment (kerwin-zk) [reply velox/docs/functions/spark/datetime.rst]: @mbasmanova done
